src/main/java/org/elasticsearch/index/mapper/internal/SourceFieldMapper.java

```diff
@@ -420,19 +429,23 @@ public XContentBuilder toXContent(XContentBuilder builder, Params params) throws
     @Override
     public void merge(Mapper mergeWith, MergeResult mergeResult) throws MergeMappingException {
         SourceFieldMapper sourceMergeWith = (SourceFieldMapper) mergeWith;
-        if (!mergeResult.simulate()) {
+        if (mergeResult.simulate()) {
+            if (this.enabled != sourceMergeWith.enabled) {
+                mergeResult.addConflict("Cannot update enabled setting for [_source]");
+            }
+            if (Arrays.equals(this.includes, sourceMergeWith.includes) == false) {
+                mergeResult.addConflict("Cannot update includes setting for [_source]");
+            }
+            if (Arrays.equals(this.excludes, sourceMergeWith.excludes) == false) {
+                mergeResult.addConflict("Cannot update excludes setting for [_source]");
+            }
```

Comment:
I was just following what I saw other mappers doing. Really I think merge should be split into two separate methods, one for validating updates, and the other for applying them?